File: qpal/src/parser/parser.js
/**
 * QPAL Pseudo-Language Parser
 * Converts pseudo-code text → structured instruction objects
 */

const VALID_COMMANDS = new Set([
  'INIT', 'H', 'X', 'CNOT', 'MEASURE',
  'ACTOR', 'ASSIGN', 'SEND', 'INTERCEPT', 'COMMENT'
]);

const BASIS_OPTIONS = new Set(['X', 'Z']);

/**
 * Parse a single line into an instruction object
 */
function parseLine(line, lineNumber) {
  const trimmed = line.trim();

  // Skip empty lines
  if (!trimmed) return null;

  // Handle comments
  if (trimmed.startsWith('#') || trimmed.startsWith('//')) {
    const text = trimmed.startsWith('//') ? trimmed.slice(2).trim() : trimmed.slice(1).trim();
    return { type: 'COMMENT', text, lineNumber };
  }

  const tokens = trimmed.split(/\s+/);
  const command = tokens[0].toUpperCase();

  if (!VALID_COMMANDS.has(command)) {
    return {
      type: 'ERROR',
      lineNumber,
      raw: trimmed,
      message: `Unknown command: "${tokens[0]}". Valid commands: ${[...VALID_COMMANDS].filter(c => c !== 'COMMENT').join(', ')}`
    };
  }

  switch (command) {
    case 'INIT': {
      if (tokens.length < 2) {
        return { type: 'ERROR', lineNumber, raw: trimmed, message: 'INIT requires a qubit name. Example: INIT q0' };
      }
      return { type: 'INIT', qubit: tokens[1].toLowerCase(), lineNumber, raw: trimmed };
    }

    case 'H': {
      if (tokens.length < 2) {
        return { type: 'ERROR', lineNumber, raw: trimmed, message: 'H (Hadamard) requires a qubit name. Example: H q0' };
      }
      return { type: 'H', qubit: tokens[1].toLowerCase(), lineNumber, raw: trimmed };
    }

    case 'X': {
      if (tokens.length < 2) {
        return { type: 'ERROR', lineNumber, raw: trimmed, message: 'X (Pauli-X / NOT gate) requires a qubit name. Example: X q0' };
      }
      return { type: 'X', qubit: tokens[1].toLowerCase(), lineNumber, raw: trimmed };
    }

    case 'CNOT': {
      if (tokens.length < 3) {
        return { type: 'ERROR', lineNumber, raw: trimmed, message: 'CNOT requires control and target qubits. Example: CNOT q0 q1' };
      }
      return {
        type: 'CNOT',
        control: tokens[1].toLowerCase(),
        target: tokens[2].toLowerCase(),
        lineNumber,
        raw: trimmed
      };
    }

    case 'MEASURE': {
      if (tokens.length < 2) {
        return { type: 'ERROR', lineNumber, raw: trimmed, message: 'MEASURE requires a qubit name. Example: MEASURE q0 or MEASURE q0 BASIS Z' };
      }
      let basis = 'Z'; // default basis
      if (tokens.length >= 4 && tokens[2].toUpperCase() === 'BASIS') {
        const basisArg = tokens[3].toUpperCase();
        if (!BASIS_OPTIONS.has(basisArg)) {
          return { type: 'ERROR', lineNumber, raw: trimmed, message: `Invalid basis "${tokens[3]}". Must be X or Z.` };
        }
        basis = basisArg;
      }
      return { type: 'MEASURE', qubit: tokens[1].toLowerCase(), basis, lineNumber, raw: trimmed };
    }

    case 'ACTOR': {
      if (tokens.length < 2) {
        return { type: 'ERROR', lineNumber, raw: trimmed, message: 'ACTOR requires a name. Example: ACTOR Alice' };
      }
      return { type: 'ACTOR', name: tokens[1], lineNumber, raw: trimmed };
    }

    case 'ASSIGN': {
      if (tokens.length < 3) {
        return { type: 'ERROR', lineNumber, raw: trimmed, message: 'ASSIGN requires qubit and actor. Example: ASSIGN q0 Alice' };
      }
      return { type: 'ASSIGN', qubit: tokens[1].toLowerCase(), actor: tokens[2], lineNumber, raw: trimmed };
    }

    case 'SEND': {
      if (tokens.length < 4) {
        return { type: 'ERROR', lineNumber, raw: trimmed, message: 'SEND requires qubit, sender and receiver. Example: SEND q0 Alice Bob' };
      }
      return {
        type: 'SEND',
        qubit: tokens[1].toLowerCase(),
        from: tokens[2],
        to: tokens[3],
        lineNumber,
        raw: trimmed
      };
    }

    case 'INTERCEPT': {
      if (tokens.length < 3) {
        return { type: 'ERROR', lineNumber, raw: trimmed, message: 'INTERCEPT requires qubit and interceptor. Example: INTERCEPT q0 Eve' };
      }
      return { type: 'INTERCEPT', qubit: tokens[1].toLowerCase(), interceptor: tokens[2], lineNumber, raw: trimmed };
    }

    default:
      return { type: 'ERROR', lineNumber, raw: trimmed, message: `Unhandled command: ${command}` };
  }
}

/**
 * Parse full program text → array of instruction objects
 */
export function parseProgram(code) {
  const lines = code.split('\n');
  const instructions = [];
  const errors = [];

  lines.forEach((line, idx) => {
    const instruction = parseLine(line, idx + 1);
    if (!instruction) return; // skip empty lines
    if (instruction.type === 'ERROR') {
      errors.push(instruction);
    }
    instructions.push(instruction);
  });

  return { instructions, errors, valid: errors.length === 0 };
}

File: backend/core/workspace/parser.py
"""QPAL pseudocode parser — Python port of the frontend pseudoParser.ts.

Converts raw QPAL custom pseudocode text into a list of WorkspaceInstruction
objects that the executor can run directly.
"""
from __future__ import annotations

import re
from typing import List

from api.schemas.workspace import WorkspaceInstruction

_COMMENT_RE = [re.compile(r"^\s*#"), re.compile(r"^\s*//")]
_QUBIT_RE = re.compile(r"^q\d+$", re.IGNORECASE)
_ACTOR_RE = re.compile(r"^[A-Za-z][A-Za-z0-9_-]*$")
_ANGLE_RE = re.compile(r"^[-+]?(\d+(\.\d+)?|\.\d+)$")
_INIT_STATES = {"0", "1", "+", "-"}

_SINGLE_QUBIT_GATES = {"H", "X", "Y", "Z", "S", "T", "SDG", "TDG", "SX", "SUPERPOSE"}
_ROTATION_GATES = {"RX", "RY", "RZ"}
_TWO_QUBIT_OPS = {"CNOT", "SWAP", "CZ", "BELL", "ENTANGLE"}


def _is_comment(line: str) -> bool:
    return any(p.search(line) for p in _COMMENT_RE)


def _validate_qubit(token: str) -> bool:
    return bool(_QUBIT_RE.match(token))


def _validate_actor(token: str) -> bool:
    return bool(_ACTOR_RE.match(token))


def _expand_macro(line: int, raw: str, opcode: str, qubits: List[str]) -> List[WorkspaceInstruction]:
    if opcode == "SUPERPOSE":
        return [
            WorkspaceInstruction(
                line=line, raw=raw, opcode="H", args=[qubits[0]],
                qubits=[qubits[0]], category="quantum",
                metadata={"expanded_from": "SUPERPOSE"},
            ),
        ]

    if opcode == "ENTANGLE":
        return [
            WorkspaceInstruction(
                line=line, raw=raw, opcode="H", args=[qubits[0]],
                qubits=[qubits[0]], category="quantum",
                metadata={"expanded_from": "ENTANGLE"},
            ),
            WorkspaceInstruction(
                line=line, raw=raw, opcode="CNOT", args=[qubits[0], qubits[1]],
                qubits=[qubits[0], qubits[1]], category="quantum",
                metadata={"expanded_from": "ENTANGLE"},
            ),
        ]

    # BELL
    return [
        WorkspaceInstruction(line=line, raw=raw, opcode="INIT", args=[qubits[0]], qubits=[qubits[0]], category="quantum", metadata={"state": "0", "expanded_from": "BELL"}),
        WorkspaceInstruction(line=line, raw=raw, opcode="INIT", args=[qubits[1]], qubits=[qubits[1]], category="quantum", metadata={"state": "0", "expanded_from": "BELL"}),
        WorkspaceInstruction(line=line, raw=raw, opcode="H", args=[qubits[0]], qubits=[qubits[0]], category="quantum", metadata={"expanded_from": "BELL"}),
        WorkspaceInstruction(line=line, raw=raw, opcode="CNOT", args=[qubits[0], qubits[1]], qubits=[qubits[0], qubits[1]], category="quantum", metadata={"expanded_from": "BELL"}),
    ]


def _parse_measure_basis(tokens: List[str]) -> tuple[str | None, int]:
    cleaned = [t.strip("[]").upper() for t in tokens]
    if not cleaned:
        return "Z", 0
    if cleaned[0] == "BASIS" and len(cleaned) > 1 and cleaned[1] in ("X", "Z"):
        return cleaned[1], 2
    if cleaned[0] in ("X", "Z"):
        return cleaned[0], 1
    return None, 0


def _invalid_instruction(line_number: int, message: str) -> ValueError:
    return ValueError(f"Line {line_number}: {message}")


def parse_pseudocode(source: str) -> List[WorkspaceInstruction]:
    """Parse QPAL pseudocode text and return a list of WorkspaceInstruction."""
    instructions: List[WorkspaceInstruction] = []

    for index, raw_line in enumerate(source.split("\n")):
        line_number = index + 1
        normalized = raw_line.replace(",", " ").strip()

        if not normalized or _is_comment(normalized):
            continue

        parts = normalized.split()
        opcode = parts[0].upper()
        rest = parts[1:]
        raw = raw_line.strip()

        # ── Annotations ──
        if opcode in ("NOTE", "LABEL"):
            text = " ".join(rest).strip()
            if not text:
                continue
            instructions.append(WorkspaceInstruction(
                line=line_number, raw=raw, opcode=opcode, args=rest,
                label=text, category="annotation",
            ))
            continue

        if opcode == "BARRIER":
            instructions.append(WorkspaceInstruction(
                line=line_number, raw=raw, opcode=opcode, category="annotation",
            ))
            continue

        if opcode == "WAIT":
            try:
                duration = float(rest[0]) if rest else 1
            except ValueError:
                raise ValueError(f"Invalid duration for WAIT: {rest[0]}")
            instructions.append(WorkspaceInstruction(
                line=line_number, raw=raw, opcode=opcode, args=rest,
                category="annotation", metadata={"duration": duration},
            ))
            continue

        # ── Actor ──
        if opcode == "ACTOR":
            if len(rest) != 1 or not _validate_actor(rest[0]):
                raise _invalid_instruction(line_number, "ACTOR expects a single valid actor name.")
            instructions.append(WorkspaceInstruction(
                line=line_number, raw=raw, opcode=opcode, args=rest,
                actors=[rest[0]], category="actor",
            ))
            continue

        # ── INIT ──
        if opcode == "INIT":
            if len(rest) < 1 or len(rest) > 2:
                raise _invalid_instruction(line_number, "INIT expects `INIT qN` or `INIT qN <state>`.")
            qubit = rest[0]
            state = rest[1] if len(rest) == 2 else "0"
            if not _validate_qubit(qubit):
                raise _invalid_instruction(line_number, f"Invalid qubit name `{qubit}` for INIT.")
            if state not in _INIT_STATES:
                raise _invalid_instruction(line_number, f"Invalid INIT state `{state}`. Expected one of {sorted(_INIT_STATES)}.")
            instructions.append(WorkspaceInstruction(
                line=line_number, raw=raw, opcode=opcode, args=rest,
                qubits=[qubit], category="quantum", metadata={"state": state},
            ))
            continue

        # ── RESET ──
        if opcode == "RESET":
            if len(rest) != 1 or not _validate_qubit(rest[0]):
                raise _invalid_instruction(line_number, "RESET expects a single valid qubit name.")
            instructions.append(WorkspaceInstruction(
                line=line_number, raw=raw, opcode=opcode, args=rest,
                qubits=[rest[0]], category="quantum",
            ))
            continue

        # ── Single-qubit gates ──
        if opcode in _SINGLE_QUBIT_GATES:
            if len(rest) != 1 or not _validate_qubit(rest[0]):
                raise _invalid_instruction(line_number, f"{opcode} expects a single valid qubit name.")
            qubit = rest[0]
            if opcode == "SUPERPOSE":
                instructions.extend(_expand_macro(line_number, raw, opcode, [qubit]))
                continue
            instructions.append(WorkspaceInstruction(
                line=line_number, raw=raw, opcode=opcode, args=rest,
                qubits=[qubit], category="quantum",
            ))
            continue

        # ── Rotation gates ──
        if opcode in _ROTATION_GATES:
            if len(rest) != 2:
                raise _invalid_instruction(line_number, f"{opcode} expects `QUBIT ANGLE`.")
            qubit, angle_tok = rest
            if not _validate_qubit(qubit):
                raise _invalid_instruction(line_number, f"Invalid qubit name `{qubit}` for {opcode}.")
            if not _ANGLE_RE.match(angle_tok):
                raise _invalid_instruction(line_number, f"Invalid angle `{angle_tok}` for {opcode}.")
            angle = float(angle_tok)
            instructions.append(WorkspaceInstruction(
                line=line_number, raw=raw, opcode=opcode, args=rest,
                qubits=[qubit], category="quantum", metadata={"angle": angle},
            ))
            continue

        # ── Two-qubit ops ──
        if opcode in _TWO_QUBIT_OPS:
            if len(rest) != 2:
                raise _invalid_instruction(line_number, f"{opcode} expects exactly two qubit operands.")
            left, right = rest
            if not _validate_qubit(left) or not _validate_qubit(right):
                raise _invalid_instruction(line_number, f"{opcode} expects valid qubit names like q0 q1.")
            if opcode in ("BELL", "ENTANGLE"):
                instructions.extend(_expand_macro(line_number, raw, opcode, [left, right]))
                continue
            instructions.append(WorkspaceInstruction(
                line=line_number, raw=raw, opcode=opcode, args=rest,
                qubits=[left, right], category="quantum",
            ))
            continue

        # ── TOFFOLI ──
        if opcode == "TOFFOLI":
            if len(rest) != 3:
                raise _invalid_instruction(line_number, "TOFFOLI expects exactly three qubit operands.")
            if not all(_validate_qubit(q) for q in rest):
                raise _invalid_instruction(line_number, "TOFFOLI expects valid qubit names like q0 q1 q2.")
            instructions.append(WorkspaceInstruction(
                line=line_number, raw=raw, opcode=opcode, args=rest,
                qubits=list(rest), category="quantum",
            ))
            continue

        # ── MEASURE ──
        if opcode == "MEASURE":
            if len(rest) < 1 or not _validate_qubit(rest[0]):
                raise _invalid_instruction(line_number, "MEASURE expects a valid qubit name.")
            qubit = rest[0]
            basis, _ = _parse_measure_basis(rest[1:])
            if basis is None:
                raise _invalid_instruction(line_number, "MEASURE basis must be X or Z when provided.")
            instructions.append(WorkspaceInstruction(
                line=line_number, raw=raw, opcode=opcode, args=rest,
                qubits=[qubit], basis=basis, category="quantum",
            ))
            continue

        # ── ASSIGN ──
        if opcode == "ASSIGN":
            if len(rest) != 2 or not _validate_qubit(rest[0]) or not _validate_actor(rest[1]):
                raise _invalid_instruction(line_number, "ASSIGN expects `ASSIGN qN ActorName`.")
            instructions.append(WorkspaceInstruction(
                line=line_number, raw=raw, opcode=opcode, args=rest,
                qubits=[rest[0]], actors=[rest[1]], category="actor",
            ))
            continue

        # ── SEND ──
        if opcode == "SEND":
            if len(rest) != 3 or not _validate_qubit(rest[0]) or not _validate_actor(rest[1]) or not _validate_actor(rest[2]):
                raise _invalid_instruction(line_number, "SEND expects `SEND qN FromActor ToActor`.")
            instructions.append(WorkspaceInstruction(
                line=line_number, raw=raw, opcode=opcode, args=rest,
                qubits=[rest[0]], actors=[rest[1], rest[2]], category="transport",
            ))
            continue

        # ── INTERCEPT ──
        if opcode == "INTERCEPT":
            if len(rest) != 2 or not _validate_qubit(rest[0]) or not _validate_actor(rest[1]):
                raise _invalid_instruction(line_number, "INTERCEPT expects `INTERCEPT qN ActorName`.")
            instructions.append(WorkspaceInstruction(
                line=line_number, raw=raw, opcode=opcode, args=rest,
                qubits=[rest[0]], actors=[rest[1]], category="transport",
            ))
            continue

        # ── Unknown opcode — strict, no silent skips ──
        raise ValueError(f"Unrecognized instruction: {opcode} on line {line_number}")

    return instructions
